src/test/groovy/unit/org/pih/warehouse/invoice/PrepaymentInvoiceServiceSpec.groovy

```diff
@@ -2,81 +2,361 @@ package unit.org.pih.warehouse.invoice
+        1        | 1.0       | 50                | 100                  | null             || 0.5                | 1
+        1        | 1.0       | 100               | 50                   | null             || 0.5                | 0.5
+        2        | 2.0       | 100               | 100                  | null             || 4                  | 4
+        2        | 2.0       | 100               | 50                   | null             || 2                  | 2
+        2        | 2.0       | 50                | 100                  | null             || 2                  | 4
+        2        | 2.0       | 50                | 50                   | null             || 1                  | 2
+        // Adjustments with a fixed amount should always have an invoiceItemAmount of adjustmentAmount * prepaymentPercent
+        1        | 1.0       | 100               | null                 | 1                || 1                  | 1
+        2        | 1.0       | 100               | null                 | 1                || 1                  | 1
+        1        | 2.0       | 100               | null                 | 1                || 1                  | 1
+        1        | 1.0       | 50                | null                 | 1                || 0.5                | 1
+        1        | 1.0       | 100               | 50                   | 1                || 1                  | 1
+        2        | 1.0       | 100               | 50                   | 1                || 1                  | 1
+        1        | 2.0       | 100               | 50                   | 1                || 1                  | 1
+        1        | 2.0       | 50                | 50                   | 1                || 0.5                | 1
+        // Adjustments with no percent or amount should always have a zero invoiceItemAmount
+        1        | 1.0       | 100               | null                 | null             || 0                  | 0
+        2        | 1.0       | 100               | null                 | null             || 0                  | 0
+        1        | 2.0       | 100               | null                 | null             || 0                  | 0
+        1        | 1.0       | 50                | null                 | null             || 0                  | 0
+    }
+
+    void 'generateInvoice should fail when prepayment invoice is missing'() {
+        given:
+        Order order = new Order(
+                invoices: [],
+        )
+
+        when:
+        service.generateInvoice(order)
+
+        then:
+        thrown(Exception)
+    }
+
+    private List<InvoiceItem> getInvoiceItemsOnOrderItems(Invoice invoice) {
+        return invoice.invoiceItems.stream()
+                .filter {ii -> ii.orderItems != null && !ii.orderItems.isEmpty()}
+                .collect(Collectors.toList())
```

Comment:
it's Java's approach for such operations, but in groovy you can just do:
```groovy
invoice.invoiceItems.findAll { it.ordersItems != null && !it.orderItems.isEmpty() }
```
btw. I assume this can't be `null`, because it's 1:N association that is by default an empty set.

Reply:
You're right. Even better would be to rely on groovy truth:
```
invoice.invoiceItems.findAll{ii -> ii.orderAdjustments}
```